File: rust/src/signal.rs
use std::mem;
use std::ptr;

use crate::error::TapestryError;

static mut GOT_SIGWINCH: bool = false;

pub fn register_sigwinch_handler() -> Result<(), TapestryError> {
    unsafe {
        let mut sa: libc::sigaction = mem::zeroed();
        sa.sa_sigaction = (handle_sigwinch as *const ()) as usize;
        libc::sigemptyset(&mut sa.sa_mask);
        sa.sa_flags = libc::SA_RESTART;

        if libc::sigaction(libc::SIGWINCH, &sa, ptr::null_mut()) < 0 {
            return Err(TapestryError::new("sigaction() failed".to_string()));
        }
    }

    eprintln!("registered sigwinch handler");
    Ok(())
}

pub fn consume_sigwinch() -> bool {
    unsafe {
        let val = GOT_SIGWINCH;
        // <--- If the value gets set to true right here, we lose it. Oh well
        GOT_SIGWINCH = false;
        val
    }
}

fn handle_sigwinch() {
    unsafe {
        GOT_SIGWINCH = true;
    }
} 

File: zig/src/signal.zig
const std = @import("std");
const posix = std.posix;

var got_sigwinch: bool = false;

pub fn register_sigwinch_handler() !void {
    posix.sigaction(
        posix.SIG.WINCH,
        &posix.Sigaction{
            .handler = .{
                .handler = handle_sigwinch,
            },
            .mask = posix.empty_sigset,
            .flags = 0,
        },
        null,
    );
}

pub fn consume_sigwinch() bool {
    const val = got_sigwinch;
    if (val) {
        got_sigwinch = false;
    }

    return val;
}

fn handle_sigwinch(_: c_int) callconv(.C) void {
    got_sigwinch = true;
}
